### PR TITLE
Ebpf disk latency

### PR DESCRIPTION
**Pull request by thiagoftsm**
##### Summary
This PR is bringing disk latency monitoring independent of filesystem used on hard disks.

##### Component Name
ebpf.plugin
##### Test Plan

1 - Edit your `ebpf.d.conf`:

```conf
[ebpf programs]
    disk = yes
```
2 - compile this branch

##### Additional Information
This PR was already tested on:

| Distribution | Kernel Version |
|--------------|----------------|
|Slackware current | 5.12.12 |
|Manjaro  | 5.10.42-1 |
|Ubuntu 18.0.4  | 5.4.114 |
|Debian 10.9  | 4.19.171 |
|Ubuntu 18.0.4  | 4.15.18 |
|Slackware current | 4.14.236 |
|CentOS 7.9  | 3.10.0-1160.25 |

### CODE REVIEW

File: collectors/ebpf.plugin/ebpf_filesystem.c [hunk 1]
```diff
@@ -114,7 +114,8 @@ static void ebpf_create_fs_charts()
                               title,
                               EBPF_COMMON_DIMENSION_CALL, family,
                               NULL, NETDATA_EBPF_CHART_TYPE_STACKED, order, ebpf_create_global_dimension,
-                              filesystem_publish_aggregated, NETDATA_FILESYSTEM_MAX_BINS);
+                              filesystem_publish_aggregated,
+                NETDATA_EBPF_HIST_MAX_BINS);
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_filesystem.c:118]: Please fix the formatting. There are 3 more similar places in this file).

File: collectors/ebpf.plugin/ebpf_disk.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+#include "libnetdata/avl/avl.h"
+#include "libnetdata/ebpf/ebpf.h"
+
+#define NETDATA_EBPF_PROC_PARTITIONS "/proc/partitions"
+
+#define NETDATA_LATENCY_DISK_SLEEP_MS 650000ULL
+
+#define NETDATA_DISK_MAX 256U
+#define NETDATA_DISK_HISTOGRAM_LENGTH (NETDATA_DISK_MAX * NETDATA_EBPF_HIST_MAX_BINS)
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_disk.h:14]: These macros are already defined in `libnetdata/ebpf/ebpf.h`.

File: libnetdata/ebpf/ebpf.c [hunk 3]
```diff
@@ -701,3 +701,114 @@ void ebpf_histogram_dimension_cleanup(char **ptr, size_t length)
...
+
+//----------------------------------------------------------------------------------------------------------------------
+
+/**
+ * Open tracepoint path
+ *
+ * @param filename    pointer to store the path
+ * @param length      file length
+ * @param subsys     is the name of your subsystem.
+ * @param eventname  is the name of the event to trace.
```
Comment (vlvkobal) [inline libnetdata/ebpf/ebpf.c:713]: Please fix the indentation of the description.

File: collectors/ebpf.plugin/ebpf_filesystem.c (hunk 1 above)
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_filesystem.c]: This was thanks `Clion` option to rename variable. :/ 
I fixed it, thank you!

File: collectors/ebpf.plugin/ebpf_disk.h (hunk 2 above)
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_disk.h]: It was removed. Thanks!

**Comment by vlvkobal** [comment]
There is something wrong with charts
image

**Comment by underhood** [comment]
seem to be getting constant 0s on ubuntu

**Comment by underhood** [comment]
fed34 VM seems to be missing the charts despite when i run the `ebpf.plugin` I can see them ... strange

File: collectors/ebpf.plugin/ebpf_disk.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+typedef struct netdata_ebpf_publish_disk {
+    netdata_ebpf_disks_t *plot;
+    struct netdata_ebpf_publish_disk *next;
+} ebpf_publish_disk_t;
+
+extern struct config disk_config;
+
+extern void *ebpf_disk_thread(void *ptr);
+
+#endif /* NETDATA_EBPF_DISK_H */
```
Comment (underhood) [inline collectors/ebpf.plugin/ebpf_disk.h:73]: \n

Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_disk.h]: Thank you I added!

**Comment by thiagoftsm** [comment]
> There is something wrong with the charts

Yes  there was, yesterday I was working to improve the original algorithm and I forgot to change a specific flag. :/ I tested on Slackware current, latest Manjaro and Ubuntu 18.04 and the problem is not happening.
@underhood I will test on Fedora to confirm.

**Comment by thiagoftsm** [comment]
@underhood I finished the tests, and the problem was also fixed on Fedora, as expected.

**Comment by vlvkobal** [comment]
`disk_latency_input.sda` now looks good, but `disk_latency_output.sda` is always zero.

**Comment by vlvkobal** [comment]
I think it's better to use `filename` instead of `text`.

**Comment by vlvkobal** [comment]
Use another variable where `text` is used as a buffer. It's misleading when you try to use the same variable with a vague name for different purposes.

**Comment by thiagoftsm** [comment]
> `disk_latency_input.sda` now looks good, but `disk_latency_output.sda` is always zero.

The charts were not wrong according benchmarks I ran with `ebpf_exporter` and `iolatency`, but I agree with you this will not give good experience for our users, so I made this [PR](https://github.com/netdata/kernel-collector/pull/238). I will make a new eBPF release before we proceed with this PR.

**Review by underhood — DISMISSED** [review]

File: web/gui/dashboard_info.js [hunk 5]
```diff
@@ -1507,6 +1507,10 @@ netdataDashboard.context = {
         height: 0.5,
         info: 'The average service time for completed I/O operations. This metric is calculated using the total busy time of the disk and the number of completed operations. If the disk is able to execute multiple parallel operations the reporting average service time will be misleading.'
     },
+    'disk.latency_io': {
+        height: 0.5,
+        info: 'Latency Output is the time it takes for an Output request to be completed. Latency is the single most important metric to focus on when it comes to storage performance, under most circumstances. For hard drives, an average latency somewhere between 10 to 20 ms can be considered acceptable. For SSD (Solid State Drives), depending on the workload it should never reach higher than 1-3 ms. In most cases, workloads will experience less than 1ms latency numbers. The dimensions refer to time intervals. This chart is based on the <a href="https://github.com/cloudflare/ebpf_exporter/blob/master/examples/bio-tracepoints.yaml" target="_blank">bio_tracepoints</a> tool of the ebpf_exporter.'
```
Comment (vlvkobal) [inline web/gui/dashboard_info.js:1512]: `Latency Output` -> `I/O latency`,
`Output request` -> `I/O request`

Comment (thiagoftsm) [reply web/gui/dashboard_info.js]: Thank you, fixed!

**Comment by vlvkobal** [comment]
Sometimes I get an error on exit. It's not critical, though.
```
2021-07-01 07:59:58: ebpf.plugin INFO  : DCSTAT KERNEL : ==4139560== Thread 17 DISK KERNEL:
==4139560== Syscall param bpf(attr->value) points to unaddressable byte(s)
==4139560==    at 0x4F3118D: syscall (in /usr/lib/libc-2.33.so)
==4139560==    by 0x144C29: sys_bpf (in /home/vlad/netdata/ebpf.plugin)
==4139560==    by 0x1459AB: bpf_map_lookup_elem (in /home/vlad/netdata/ebpf.plugin)
==4139560==    by 0x114E56: read_hard_disk_tables (ebpf_disk.c:509)
==4139560==    by 0x114F1F: ebpf_disk_read_hash (ebpf_disk.c:577)
==4139560==    by 0x140BD0: thread_start (threads.c:184)
==4139560==    by 0x4E20258: start_thread (in /usr/lib/libpthread-2.33.so)
==4139560==    by 0x4F365E2: clone (in /usr/lib/libc-2.33.so)
==4139560==  Address 0x19dd27a0 is 0 bytes inside a block of size 64 free'd
==4139560==    at 0x484118B: free (vg_replace_malloc.c:755)
==4139560==    by 0x131075: freez (libnetdata.c:173)
==4139560==    by 0x11511B: ebpf_disk_cleanup (ebpf_disk.c:438)
==4139560==    by 0x11571A: ebpf_disk_thread (ebpf_disk.c:816)
==4139560==    by 0x140BD0: thread_start (threads.c:184)
==4139560==    by 0x4E20258: start_thread (in /usr/lib/libpthread-2.33.so)
==4139560==    by 0x4F365E2: clone (in /usr/lib/libc-2.33.so)
==4139560==  Block was alloc'd at
==4139560==    at 0x48435FF: calloc (vg_replace_malloc.c:1117)
==4139560==    by 0x131111: callocz (libnetdata.c:183)
==4139560==    by 0x1155D0: disk_collector (ebpf_disk.c:699)
==4139560==    by 0x11587D: ebpf_disk_thread (ebpf_disk.c:813)
==4139560==    by 0x140BD0: thread_start (threads.c:184)
==4139560==    by 0x4E20258: start_thread (in /usr/lib/libpthread-2.33.so)
==4139560==    by 0x4F365E2: clone (in /usr/lib/libc-2.33.so)
==4139560==
==4139560== Invalid read of size 8
==4139560==    at 0x114DF4: read_hard_disk_tables (ebpf_disk.c:543)
==4139560==    by 0x114F1F: ebpf_disk_read_hash (ebpf_disk.c:577)
==4139560==    by 0x140BD0: thread_start (threads.c:184)
==4139560==    by 0x4E20258: start_thread (in /usr/lib/libpthread-2.33.so)
==4139560==    by 0x4F365E2: clone (in /usr/lib/libc-2.33.so)
==4139560==  Address 0x19dd27a8 is 8 bytes inside a block of size 64 free'd
==4139560==    at 0x484118B: free (vg_replace_malloc.c:755)
==4139560==    by 0x131075: freez (libnetdata.c:173)
==4139560==    by 0x11511B: ebpf_disk_cleanup (ebpf_disk.c:438)
==4139560==    by 0x11571A: ebpf_disk_thread (ebpf_disk.c:816)
==4139560==    by 0x140BD0: thread_start (threads.c:184)
==4139560==    by 0x4E20258: start_thread (in /usr/lib/libpthread-2.33.so)
==4139560==    by 0x4F365E2: clone (in /usr/lib/libc-2.33.so)
==4139560==  Block was alloc'd at
==4139560==    at 0x48435FF: calloc (vg_replace_malloc.c:1117)
==4139560==    by 0x131111: callocz (libnetdata.c:183)
==4139560==    by 0x1155D0: disk_collector (ebpf_disk.c:699)
==4139560==    by 0x11587D: ebpf_disk_thread (ebpf_disk.c:813)
==4139560==    by 0x140BD0: thread_start (threads.c:184)
==4139560==    by 0x4E20258: start_thread (in /usr/lib/libpthread-2.33.so)
==4139560==    by 0x4F365E2: clone (in /usr/lib/libc-2.33.so)
==4139560==
thread with task id 4139578 finished
```

File: collectors/ebpf.plugin/ebpf_disk.c [hunk 6]
```diff
@@ -0,0 +1,819 @@
...
+ *
+ * Make Hard disk charts and fill chart name
+ */
+static void ebpf_create_hd_charts(netdata_ebpf_disks_t *w)
+{
+    char title[256];
+    int order = NETDATA_CHART_PRIO_DISK_LATENCY;
+    char *family = w->family;
+
+    snprintf(title, 255, "Disk latency %s.", family);
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_disk.c:612]: We don't need the disk name in the chart title.

File: web/gui/dashboard_info.js [hunk 7]
```diff
@@ -1507,6 +1507,10 @@ netdataDashboard.context = {
         height: 0.5,
         info: 'The average service time for completed I/O operations. This metric is calculated using the total busy time of the disk and the number of completed operations. If the disk is able to execute multiple parallel operations the reporting average service time will be misleading.'
     },
+    'disk.latency_io': {
+        height: 0.5,
+        info: 'Latency I/O is the time it takes for an I/O request to be completed. Latency is the single most important metric to focus on when it comes to storage performance, under most circumstances. For hard drives, an average latency somewhere between 10 to 20 ms can be considered acceptable. For SSD (Solid State Drives), depending on the workload it should never reach higher than 1-3 ms. In most cases, workloads will experience less than 1ms latency numbers. The dimensions refer to time intervals. This chart is based on the <a href="https://github.com/cloudflare/ebpf_exporter/blob/master/examples/bio-tracepoints.yaml" target="_blank">bio_tracepoints</a> tool of the ebpf_exporter.'
```
Comment (vlvkobal) [inline web/gui/dashboard_info.js:1512]:
```suggestion
        info: 'Disk I/O latency is the time it takes for an I/O request to be completed. Latency is the single most important metric to focus on when it comes to storage performance, under most circumstances. For hard drives, an average latency somewhere between 10 to 20 ms can be considered acceptable. For SSD (Solid State Drives), depending on the workload it should never reach higher than 1-3 ms. In most cases, workloads will experience less than 1ms latency numbers. The dimensions refer to time intervals. This chart is based on the <a href="https://github.com/cloudflare/ebpf_exporter/blob/master/examples/bio-tracepoints.yaml" target="_blank">bio_tracepoints</a> tool of the ebpf_exporter.'
```

File: collectors/ebpf.plugin/ebpf_disk.c [hunk 8]
```diff
@@ -0,0 +1,819 @@
...
+
+    snprintfz(filename, FILENAME_MAX, "%s/%s/%s/size", path, disk, name);
+    ebpf_parse_size(w, filename);
+}
+
+// Decode function extracted from: https://elixir.bootlin.com/linux/v5.10.8/source/include/linux/kdev_t.h#L46
+/**
+ * New encode dev
+ *
+ * New enconde algorithm
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_disk.c:205]: Please fix the description: wrong link, confusing description.

File: collectors/ebpf.plugin/ebpf_disk.c [hunk 9]
```diff
@@ -0,0 +1,819 @@
...
+}
+
+/**
+ * Update listen table
+ *
+ * Update link list when it is necessary.
+ *
+ * @param name  the    disk name
+ * @param major major  disk identifier
+ * @param minor minor  disk identifier
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_disk.c:246]: Wrong indentation.

File: collectors/ebpf.plugin/ebpf_disk.c [hunk 10]
```diff
@@ -0,0 +1,819 @@
...
+
+#ifdef NETDATA_INTERNAL_CHECKS
+    info("The Latency is monitoring the hard disk %s (Major = %d, Minor = %d, Device = %u)", name, major, minor,w->dev);
+#endif
+
+    w->flags |= NETDATA_DISK_IS_HERE;
+}
+
+/**
+ *  Read Local Ports
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_disk.c:316]: Ports?

**Comment by vlvkobal** [comment]
You are using the major/minor numbers for identification, have you considered possible problems like in #10841?

File: collectors/ebpf.plugin/ebpf_disk.c (hunk 10 above)
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_disk.c]: I fixed the minor problems with the last commits now I will work with trace reported.

**Comment by thiagoftsm** [comment]
> Sometimes I get an error on exit. It's not critical, though.
> 

After the last commit, neither `valgrind 3.15` nor `valgrind 3.17` reported problems. I did not observed problems with `libasan` too.

**Comment by thiagoftsm** [comment]
> You are using the major/minor numbers for identification, have you considered possible problems like in #10841?

As we discussed on Slack, we are extracting values direct from kernel variables. I think we can move forward without concerns with specific drivers. This plugin will also receive additional features, so we can bring adjusts for specific kernel drives in the future.

File: collectors/ebpf.plugin/ebpf_disk.c [hunk 11]
```diff
@@ -0,0 +1,837 @@
...
+static void ebpf_create_hd_charts(netdata_ebpf_disks_t *w)
+{
+    int order = NETDATA_CHART_PRIO_DISK_LATENCY;
+    char *family = w->family;
+
+    w->histogram.name = strdupz("disk_latency_io");
+    w->histogram.title = NULL;
+    w->histogram.order = order;
+
+    ebpf_create_chart(w->histogram.name, family, "Disk latency.", EBPF_COMMON_DIMENSION_CALL,
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_disk.c:635]:
```suggestion
    ebpf_create_chart(w->histogram.name, family, "Disk latency", EBPF_COMMON_DIMENSION_CALL,
```

Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_disk.c]: Thank you Vlad fixed!

**Review by vlvkobal — APPROVED** [review]

**Review by underhood — APPROVED** [review]

**Pull request merged**